### ISSUE TITLE
Bug: [no-confusing-void-expression] Invalid behaviour with generic function

### ISSUE THREAD

**Issue by sviat9440** [typescript-eslint/typescript-eslint#6277]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=GYVwdgxgLglg9mABFApgZygHgCoD4AUAbgIYA2AXItgJSXaIDeAUIogPRuIB0PLiATiigh+SEqQDcTAL5MmEBGjikUXUnADm+VBnwByHVD3VqUhWCUq1m7eij4wIUqRNnFy1eq2H84ACYowDBgKH6uTEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90URNGgdokcGAC+IPUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
function test<T>(val: T): T {
  // ...
  return val;
}

console.log(test('test'));
console.log(test(null));
console.log(test(undefined));
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-confusing-void-expression": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No errors

### Actual Result

Placing a void expression inside another expression is forbidden. Move it to its own statement instead. 8:13 - 8:28

**Comment by bradzacher** [comment]
This is working as expected - the rule checks for "void-like" values - not specifically just `void`. `undefined` is classed by TS as a "void-like", so the rule matches this case.

You'll notice that if you remove the generic and just make the function specifically return `undefined` that the same case will be reported.
